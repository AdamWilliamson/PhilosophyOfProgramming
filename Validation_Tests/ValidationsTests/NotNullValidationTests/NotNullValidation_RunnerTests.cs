using ApprovalTests;
using FluentAssertions;
using System.Threading.Tasks;
using Validations;
using Validations_Tests.TestHelpers;
using Validations_Tests.ValidationsTests.GenericTestableObjects;
using Xunit;

namespace Validations_Tests.ValidationsTests;

public class NotNull_NoError_TestingValidator : AbstractValidator<NullAllFieldTypesDto>
{
    public NotNull_NoError_TestingValidator()
    {
        Describe(x => x.Integer).NotNull();
        Describe(x => x.String).NotNull();
        Describe(x => x.Decimal).NotNull();
        Describe(x => x.Double).NotNull();
        Describe(x => x.Short).NotNull();
        Describe(x => x.Long).NotNull();
        Describe(x => x.TwoComponentTuple).NotNull();
        Describe(x => x.TwoComponentNewTupple).NotNull();
        Describe(x => x.AllFieldTypesList).NotNull();
        Describe(x => x.AllFieldTypesLinkedList).NotNull();
        Describe(x => x.AllFieldTypesIEnumerable).NotNull();
        Describe(x => x.AllFieldTypesDictionary).NotNull();
        Describe(x => x.Struct).NotNull();
    }
}

public class NotNullValidation_RunnerTests
{
    [Fact]
    public async Task GivenAValidator_WithNoErrors_ThenNoErrorsAreProduced()
    {
        // Arrange
        var runner = ValidationRunnerHelper.BasicRunnerSetup(new NotNull_NoError_TestingValidator());

        // Act
        var validationResult = await runner.Validate(new NullAllFieldTypesDto());

        // Assert
        validationResult.Results.Should().BeEmpty();
    }

    [Fact]
    public async Task GivenAValidator_WithErrors_ThenEveryFieldHasErrors()
    {
        // Arrange
        var runner = ValidationRunnerHelper.BasicRunnerSetup(new NotNull_NoError_TestingValidator());

        // Act
        var validationResult = await runner.Validate(new NullAllFieldTypesDto()
        {
            Integer = null,
            String = null,
            Decimal = null,
            Double = null,
            Short = null,
            Long = null,
            TwoComponentTuple = null,
            TwoComponentNewTupple = null,
            AllFieldTypesList = null,
            AllFieldTypesLinkedList = null,
            AllFieldTypesIEnumerable = null,
            AllFieldTypesDictionary = null,
            Struct = null,
        });
        var json = JsonConverter.ToJson(validationResult);

        // Assert
        validationResult.Results.Should().HaveCount(13);
        Approvals.VerifyJson(json);
    }

    [Fact]
    public void GivenAValidator_WhenDescribing_ThenEveryFieldHasDescriptions()
    {
        // Arrange
        var runner = ValidationRunnerHelper.BasicRunnerSetup(new EqualTo_NoError_TestingValidator());

        // Act
        var description = runner.Describe();
        var json = JsonConverter.ToJson(description);

        // Assert
        description.Results.Should().HaveCount(7);
        Approvals.VerifyJson(json);
    }
}
