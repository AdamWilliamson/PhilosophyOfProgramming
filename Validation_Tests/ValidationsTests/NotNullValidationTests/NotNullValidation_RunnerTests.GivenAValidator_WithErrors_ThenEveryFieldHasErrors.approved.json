{
    "results": [
        {
            "property": "Integer",
            "outcomes": [
                {
                    "isFailed": true,
                    "isSucceeded": false,
                    "propertyName": "Integer",
                    "message": "Is not null",
                    "group": null
                }
            ]
        },
        {
            "property": "String",
            "outcomes": [
                {
                    "isFailed": true,
                    "isSucceeded": false,
                    "propertyName": "String",
                    "message": "Is not null",
                    "group": null
                }
            ]
        },
        {
            "property": "Decimal",
            "outcomes": [
                {
                    "isFailed": true,
                    "isSucceeded": false,
                    "propertyName": "Decimal",
                    "message": "Is not null",
                    "group": null
                }
            ]
        },
        {
            "property": "Double",
            "outcomes": [
                {
                    "isFailed": true,
                    "isSucceeded": false,
                    "propertyName": "Double",
                    "message": "Is not null",
                    "group": null
                }
            ]
        },
        {
            "property": "Short",
            "outcomes": [
                {
                    "isFailed": true,
                    "isSucceeded": false,
                    "propertyName": "Short",
                    "message": "Is not null",
                    "group": null
                }
            ]
        },
        {
            "property": "Long",
            "outcomes": [
                {
                    "isFailed": true,
                    "isSucceeded": false,
                    "propertyName": "Long",
                    "message": "Is not null",
                    "group": null
                }
            ]
        },
        {
            "property": "TwoComponentTuple",
            "outcomes": [
                {
                    "isFailed": true,
                    "isSucceeded": false,
                    "propertyName": "TwoComponentTuple",
                    "message": "Is not null",
                    "group": null
                }
            ]
        },
        {
            "property": "TwoComponentNewTupple",
            "outcomes": [
                {
                    "isFailed": true,
                    "isSucceeded": false,
                    "propertyName": "TwoComponentNewTupple",
                    "message": "Is not null",
                    "group": null
                }
            ]
        },
        {
            "property": "AllFieldTypesList",
            "outcomes": [
                {
                    "isFailed": true,
                    "isSucceeded": false,
                    "propertyName": "AllFieldTypesList",
                    "message": "Is not null",
                    "group": null
                }
            ]
        },
        {
            "property": "AllFieldTypesLinkedList",
            "outcomes": [
                {
                    "isFailed": true,
                    "isSucceeded": false,
                    "propertyName": "AllFieldTypesLinkedList",
                    "message": "Is not null",
                    "group": null
                }
            ]
        },
        {
            "property": "AllFieldTypesIEnumerable",
            "outcomes": [
                {
                    "isFailed": true,
                    "isSucceeded": false,
                    "propertyName": "AllFieldTypesIEnumerable",
                    "message": "Is not null",
                    "group": null
                }
            ]
        },
        {
            "property": "AllFieldTypesDictionary",
            "outcomes": [
                {
                    "isFailed": true,
                    "isSucceeded": false,
                    "propertyName": "AllFieldTypesDictionary",
                    "message": "Is not null",
                    "group": null
                }
            ]
        },
        {
            "property": "Struct",
            "outcomes": [
                {
                    "isFailed": true,
                    "isSucceeded": false,
                    "propertyName": "Struct",
                    "message": "Is not null",
                    "group": null
                }
            ]
        }
    ]
}